### PR TITLE
fix: on primary action callback close msgprint

### PR DESCRIPTION
**Pull request by vishdha**
Task ID: https://bloomstack.com/desk#Form/Task/TASK-2020-00117

### CODE REVIEW

File: frappe/public/js/frappe/ui/messages.js [hunk 1]
```diff
@@ -139,6 +139,10 @@ frappe.msgprint = function(msg, title) {
 					method: data.primary_action.server_action,
 					args: {
 						args: data.primary_action.args
+					},
+					callback: (r) => {
+						console.log("sucess")
```
Comment (Alchez) [inline frappe/public/js/frappe/ui/messages.js]: Remove the console.log and put it upstream, see what they think

File: frappe/public/js/frappe/ui/messages.js [hunk 2]
```diff
@@ -139,6 +139,10 @@ frappe.msgprint = function(msg, title) {
 					method: data.primary_action.server_action,
 					args: {
 						args: data.primary_action.args
+					},
+					callback: (r) => {
+						// close msgprint dialog box on primary action
+						frappe.hide_msgprint();
```
Comment (Alchez) [inline frappe/public/js/frappe/ui/messages.js]: To get more dev control, I think you should also add a `callback` property in `data.primary_action`. That way, we can do whatever we want during callback. This is too restrictive.

**Comment by vishdha** [comment]
upstream: https://github.com/frappe/frappe/pull/10693

File: frappe/public/js/frappe/ui/messages.js [hunk 3]
```diff
@@ -139,6 +139,20 @@ frappe.msgprint = function(msg, title) {
...
 						args: data.primary_action.args
+					},
+					callback: (r) => {
+						// close msgprint dialog box on primary action
+						// frappe.hide_msgprint();
+						if (r) {
+							frappe.run_serially([
+								() => frappe.hide_msgprint(),
+								() => frappe.timeout(.5),
+								() => frappe.msgprint(r.message, title="sucess")
```
Comment (Alchez) [inline frappe/public/js/frappe/ui/messages.js]:
```suggestion
								() => frappe.msgprint(r.message, title="Success")
```

File: frappe/public/js/frappe/ui/messages.js [hunk 4]
```diff
@@ -139,6 +139,20 @@ frappe.msgprint = function(msg, title) {
 					method: data.primary_action.server_action,
 					args: {
 						args: data.primary_action.args
+					},
+					callback: (r) => {
+						// close msgprint dialog box on primary action
+						// frappe.hide_msgprint();
+						if (r) {
```
Comment (Alchez) [inline frappe/public/js/frappe/ui/messages.js]:
```suggestion
						if (r && r.message) {
```

**Review by Alchez — CHANGES_REQUESTED** [review]

**Review by Alchez — DISMISSED** [review]

File: frappe/__init__.py [hunk 5]
```diff
@@ -344,6 +344,9 @@ def _raise_exception():
 	if primary_action:
 		out.primary_action = primary_action
 
+	if hide_on_success:
+		out.hide_on_success = 1
```
Comment (Alchez) [inline frappe/__init__.py]: The `hide_on_success` is contextual to the primary action, it should be a part of that object

File: frappe/public/js/frappe/ui/messages.js [hunk 6]
```diff
@@ -139,6 +139,19 @@ frappe.msgprint = function(msg, title) {
 					method: data.primary_action.server_action,
 					args: {
 						args: data.primary_action.args
+					},
+					freeze: true,
+					callback: (r) => {
+						if (r && r.message) {
+							if (data.hide_on_success) {
```
Comment (Alchez) [inline frappe/public/js/frappe/ui/messages.js]: Let's keep it simple, see if this works

```
if (data.primary_action.hide_on_success) {
	hide_msgprint();
}

if (r && r.message) {
	frappe.msgprint();
}
```

**Review by Alchez — CHANGES_REQUESTED** [review]

**Comment by vishdha** [comment]
upstream: https://github.com/frappe/frappe/pull/10729

**Review by Alchez — APPROVED** [review]

**Pull request merged**